### PR TITLE
(feat) Allow fully "controlled" range and multiple modes

### PR DESCRIPTION
**Pull request by tjfred35**
I’m pretty sure this addresses the feature request here:

https://github.com/gpbl/react-day-picker/issues/1245

This change is to allow `selected` prop to be used for all the modes as a way for the state to be externally managed (“controlled”). This allows users to manipulate this state “without user interaction”.

Additionally, I’m hoping we can rename mode="uncontrolled" to mode="custom" -- this seems to better express how the documentation already refers to this mode, and doesn’t get mixed up with the other common definition in react component libraries like downshift and even elsewhere in react-day-picker:

- controlled means the user passes in the state
- uncontrolled means the component manages the state for the user, but the user can pass in some default state

This is the definition that react day picker uses for months. From the docs:

> To control the current month, use `month` (as opposed to `defaultMonth`) and `onMonthChange` to handle the current month.

I'm still looking into the implementation of this controlled pattern for month (useNavigationState.ts) -- it's a little different from the controlled implementation I’ve written for `selected`, so let's discuss which one we prefer. I will look into the downshift implementation too.

 And I haven't implemented this fully yet -- need to implement this same pattern for `mode='single'` for completeness.

### CODE REVIEW

**Comment by tjfred35** [comment]
@gpbl Ok this is ready for a review -- I've looked at downshift implementation of its "controlled" props:

https://github.com/downshift-js/downshift/tree/a6bcd2247f4757b9a6c4077fdd9cf07554049054/src/hooks/useSelect#control-props

It's much more complicated, but it ultimately boils down to the same thing -- when the user passes in the `selectedItem` and other props, it uses those in lieu of its own internal state.

**Review by gpbl — COMMENTED** [review]
Thanks for your work! That `useSelect` in each provider is a red flag. I think we need to review or refactor something first – let keep this on hold while I review the codebase :)

**Comment by tjfred35** [comment]
The dynamic resolution of types based on the value of mode is really interesting, but also complex enough that it’s beyond my technical ability to consolidate these three methods so I’m interested to see if you can come up with a solution

**Comment by tjfred35** [comment]
Also, I just realized that the examples I wrote for the documentation were broken. I just fixed them.

**Comment by gpbl** [comment]
> The dynamic resolution of types based on the value of mode is really interesting, but also complex enough that it’s beyond my technical ability

Well that's not good 😅 Let's block for a moment this and review the architecture.

File: packages/react-day-picker/src/contexts/SelectMultiple/SelectMultipleProvider.tsx
Comment (tjfred35) [inline packages/react-day-picker/src/contexts/SelectMultiple/SelectMultipleProvider.tsx:116]: I’m going to try something like this:
```
type ControllableProps<T> {
  prop?: T
  defaultProp?: T
}

…
const selectedProps: ControllableProps<Date[]> = isDayPickerMultiple(initialProps)
  ? { prop: selected, defaultProp: defaultSelected }
  : { }
  
const [selectedDays, setSelectedDays] = useControllablePropsState(selectedProps)
```

**Comment by tjfred35** [comment]
@gpbl Ok, I was able to extract a hook that encapsulates the logic that was previously copy / pasted across three components. Does this unblock this from moving forward?

File: packages/react-day-picker/src/types/DayPickerBase.ts
Comment (gpbl) [inline packages/react-day-picker/src/types/DayPickerBase.ts:196]: To investigate - what should be the default mode?

**Review by gpbl — APPROVED** [review]
Thanks I could check this out better, looks like it is laying out a better foundation.

I have some feedback for the `useControllableProp` hook. Will open a PR soon.

**Pull request merged**
